src/main/scala/io/iohk/ethereum/domain/BlockHeader.scala

```diff
@@ -54,43 +54,69 @@ case class BlockHeader(
     val rlpEncoded = blockHeader.toRLPEncodable match {
-      case rlpList: RLPList if blockHeader.treasuryOptOut.isEmpty =>
-        // Pre ECIP1098 block
-        RLPList(rlpList.items.dropRight(2): _*)
+      case rlpList: RLPList if blockHeader.checkpoint.isDefined =>
+        // post ECIP1098 & ECIP1097 block
+        val rlpItemsWithoutNonce = rlpList.items.dropRight(4) ++ rlpList.items.takeRight(2)
+        RLPList(rlpItemsWithoutNonce: _*)
 
       case rlpList: RLPList if blockHeader.treasuryOptOut.isDefined =>
-        // Post ECIP1098 block
+        // Post ECIP1098 block without checkpoint
         val rlpItemsWithoutNonce = rlpList.items.dropRight(3) :+ rlpList.items.last
         RLPList(rlpItemsWithoutNonce: _*)
 
+      case rlpList: RLPList if blockHeader.treasuryOptOut.isEmpty =>
+        // Pre ECIP1098 & ECIP1097 block
+        RLPList(rlpList.items.dropRight(2): _*)
+
       case _ => throw new Exception("BlockHeader cannot be encoded without nonce and mixHash")
     }
     rlpEncode(rlpEncoded)
   }
 
   implicit class BlockHeaderEnc(blockHeader: BlockHeader) extends RLPSerializable {
+    private def encodeOptOut(definedOptOut: Boolean) = {
+      val encodedOptOut = if(definedOptOut) 1 else 0
+      RLPList(encodedOptOut)
+    }
     override def toRLPEncodable: RLPEncodeable = {
       import blockHeader._
-      treasuryOptOut match {
-        case Some(definedOptOut) =>
-          // Post ECIP1098 block, whole block is encoded
-          val encodedOptOut = if(definedOptOut) 1 else 0
+      (treasuryOptOut, checkpoint) match {
+        case (Some(definedOptOut), Some(_)) =>
+          // Post ECIP1098 & ECIP1097 block, block with treasury enabled and checkpoint is encoded
+          RLPList(parentHash, ommersHash, beneficiary, stateRoot, transactionsRoot, receiptsRoot,
+            logsBloom, difficulty, number, gasLimit, gasUsed, unixTimestamp, extraData, mixHash, nonce, encodeOptOut(definedOptOut), checkpoint)
 
+        case (Some(definedOptOut), None) =>
+          // Post ECIP1098 block, Pre ECIP1097 or without checkpoint, block with treasury enabled is encoded
           RLPList(parentHash, ommersHash, beneficiary, stateRoot, transactionsRoot, receiptsRoot,
-            logsBloom, difficulty, number, gasLimit, gasUsed, unixTimestamp, extraData, mixHash, nonce, RLPList(encodedOptOut))
+            logsBloom, difficulty, number, gasLimit, gasUsed, unixTimestamp, extraData, mixHash, nonce, encodeOptOut(definedOptOut))
 
-        case None =>
-          // Pre ECIP1098 block, encoding works as if optOut field wasn't defined for backwards compatibility
+        case (None, Some(_)) =>
+          // Post ECIP1097 block with checkpoint, treasury disabled, block with checkpoint is encoded
+          RLPList(parentHash, ommersHash, beneficiary, stateRoot, transactionsRoot, receiptsRoot,
+            logsBloom, difficulty, number, gasLimit, gasUsed, unixTimestamp, extraData, mixHash, nonce, RLPList(), checkpoint)
```

Comment:
It's a bit awkward.
What if we had a single optional field interpreted as "block header extensions"? It would be always present after ECIP-X. Then we could explicitly state optionality of the new extensions in all cases, and it would be safe for any future extensions as well.

Reply:
do you mean one additional field for BlockHeader where we will pack all extensions ? I'm ok with that, but as it will touch also treasury opt, IMHO it should be done in different PR as it will add a lot of noise to the PR.

Reply:
Yes, definitely. This would be an important breaking change, so I'd like to consult that with others and plan it as proper a task. 

I'm also not sure if this is the best we can do. @ntallar WDYT?

Reply:
IMHO it would be also good to do something similar in BlockchainConfig

Reply:
Wouldn't a block header extensions field only move this complexity there, with the same matching on it's size as here? Why would that simplify any future extensions?

The only thing I think would simplify quite a lot the enconding/decoding design is having a block version field as with bitcoin (we could add it as the optional 17th field). Then:
- Version 1: no version field, traditional ETC block
- Version 2: treasuryOptOut as 18th field + checkpoint as 19th field
- ...

The spec didn't include that as I thought it might be an overkill, I haven't heard of future block headers changes for now

(I would also add any changes resulting from this discussion in a separate PR)

Reply:
> Why would that simplify any future extensions?

Not necessarily simplify, but we would always have explicitness regarding optionality. Anyway, I do like the version field idea better. I think this is the way to go 💯